### PR TITLE
gh-117323: Make `cell` thread-safe in free-threaded builds

### PR DESCRIPTION
**Pull request by colesbury**
Use critical sections to lock around accesses to cell contents. The critical sections are no-ops in the default (with GIL) build.

* Issue: gh-117323

### CODE REVIEW

File: Include/internal/pycore_cell.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+// Sets the cell contents to `value` and return previous contents. Steals a
+// reference to `value`.
+static inline PyObject *
+PyCell_SwapTakeRef(PyCellObject *cell, PyObject *value)
+{
+    PyObject *old_value;
+    Py_BEGIN_CRITICAL_SECTION(cell);
```
Comment (Fidget-Spinner) [inline Include/internal/pycore_cell.h:20]: Just checking so I understand correctly: this also implies GC triggered by another thread cannot happen in between these two sections right?

File: Tools/cases_generator/analyzer.py [hunk 2]
```diff
@@ -522,6 +522,9 @@ def compute_properties(op: parser.InstDef) -> Properties:
         variable_used(op, "PyCell_New")
         or variable_used(op, "PyCell_GET")
         or variable_used(op, "PyCell_SET")
```
Comment (brandtbucher) [inline Tools/cases_generator/analyzer.py]: Not a huge deal, but I don't think these are needed (and they shouldn't ever be added back).

```suggestion
```

**Review by brandtbucher — APPROVED** [review]
Thanks for keeping the JIT working. Looks good, just one observation:

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Just one clarification.

File: Include/internal/pycore_cell.h (hunk 1 above)
Comment (colesbury) [reply Include/internal/pycore_cell.h:20]: `Py_BEGIN_CRITICAL_SECTION()` is not directly related to GC. It locks the per-object lock on `cell`. There's a more detail description in the file comments: 
https://github.com/python/cpython/blob/main/Include/internal/pycore_critical_section.h

However, you are right that a GC cannot happen between these two calls *in this case*. The two lines below can't trigger a GC, nor do they release the GIL (or "detach" in the free-threaded build), so another thread can't perform a GC either.

**Pull request merged**